### PR TITLE
feat: contact dedup and merge

### LINKED ISSUE #42: Implement contact.merge skill (consolidate duplicate contacts)
## Background

Spec 09 describes a `contact.merge` skill that consolidates two contacts into one — merging channel identities, auth overrides, and KG links. Deferred from Phase B because:

1. We don't yet have enough contacts to produce real duplicates
2. The KG entity linking makes merge non-trivial (which entity node survives?)
3. The authorization override merge has edge cases (conflicting grants/denials)

## What merge needs to handle

- Consolidate all channel identities from both contacts onto the survivor
- Merge auth overrides (conflict resolution: explicit deny wins over grant?)
- Re-link KG edges from the deleted contact's entity to the survivor's entity
- Update working memory references (conversation history that mentions the old contact)
- Audit log entry for the merge (who merged, what was combined)
- CEO confirmation before executing

## Dependencies

- Richer KG with entity deletion support
- Auth overrides system (Phase B)

## Context

Deferred during Phase B planning. The skill, merge logic, and conflict resolution should be designed when we have real-world duplicate contacts to test against.

### PR DESCRIPTION
**Pull request by josephfung**
## Summary

- **Deterministic deduplication** (`DedupService`): Jaro-Winkler name similarity + exact channel identifier overlap + 3-char name-prefix blocking. Scores ≥ 0.9 = `certain`; 0.7–0.9 = `probable`.
- **On-creation hook**: `createContact()` fires a fire-and-forget background check (`setImmediate`) and publishes `contact.duplicate_detected` on the bus when a match is found.
- **Golden record merge**: `ContactService.mergeContacts()` computes survivorship (most-recent-wins for scalars, concat for notes, most-restrictive for status), re-points all identities and auth overrides to the primary, then deletes the secondary.
- **KG node merge**: `EntityMemory.mergeEntities()` handles scalar property merge and fact migration (Phase 1). Edge re-pointing and secondary node deletion are Phase 2 (`@TODO`).
- **New skills**: `contact-find-duplicates` (action_risk: none) and `contact-merge` (action_risk: low, elevated caller required, dry_run default true).
- **Coordinator updated** with dedup workflow guidance and both skills pinned.
- **Bus events**: `contact.duplicate_detected` and `contact.merged` — reason strings are privacy-safe (no PII/identifier values).

Closes #42.

**Not addressed in this PR (noted in code with `@TODO`):**
- KG Phase 2: re-pointing relationship edges from secondary to primary and deleting the secondary node — the secondary KG node remains as a dangling node after merge.
- Working memory references: out of scope per the design spec; a separate pass would be needed to rewrite any working memory entries referencing the secondary contact ID.

## Test plan

- [x] Unit tests: `DedupService` (Jaro-Winkler scoring, blocking, channel overlap, three-way overlap edge cases)
- [x] Unit tests: `EntityMemory.mergeEntities()` (primary wins on tie, secondary wins when newer, fact migration)
- [x] Unit tests: `ContactService.mergeContacts()` (full merge, dry_run, auth override consolidation)
- [x] Unit tests: `findDuplicates()` (shared email → certain pair)
- [x] Unit tests: `contact-find-duplicates` handler (score rounding, min_confidence filter)
- [x] Unit tests: `contact-merge` handler (dry_run, UUID validation, same-ID rejection, not-found errors)
- [x] All 771 tests passing, 16 skipped

___

## **CodeAnt-AI Description**
**Add duplicate contact review and merge flows**

### What Changed
- Added a duplicate scan that returns ranked contact pairs, with confidence levels and a simple filter for review sessions
- Added a contact merge flow with a preview mode, then live merge that keeps the primary contact, combines identities and permissions, and removes the secondary contact
- New contacts can now trigger a background duplicate check and emit duplicate/merge notifications without blocking contact creation
- Duplicate matching now catches shared channel identities and name variants, including cases with multiple contacts sharing the same identifier
- Contact and knowledge graph merge behavior now preserves the newer record when needed, keeps notes, and carries over stored facts

### Impact
`✅ Faster duplicate review`
`✅ Fewer duplicate contacts`
`✅ Clearer merge previews`
<details>
<summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

### CODE REVIEW

File: tests/unit/contacts/contact-service.test.ts [hunk 1]
```diff
@@ -340,4 +341,284 @@ describe('ContactService', () => {
...
+    });
+
+    // Make secondary's timestamp strictly newer than primary's.
+    // InMemoryBackend.getNode() returns the node object by reference (no copy),
+    // so mutating temporal.lastConfirmedAt here updates the stored node in-place.
+    // This is intentionally test-only; production code should never bypass the store API.
+    // @TODO: Once a store-level updateNode overload that accepts a full temporal object
+    // is added, switch this to use that API so the test isn't relying on reference identity.
+    const secondaryNode = await entityMemory.getEntity(secondary.id);
+    secondaryNode!.temporal.lastConfirmedAt = new Date(Date.now() + 10_000);
```
[the remark this answers is not in the document]
Comment (josephfung) [reply tests/unit/contacts/contact-service.test.ts]: Fixed in e72261b — added  before the mutation.

File: src/contacts/dedup-service.ts [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+  if (parts.length >= 2) {
+    // Last-first variant
+    variants.push([parts[parts.length - 1], ...parts.slice(0, -1)].join(' '));
+    // Initial variant (first initial + rest)
+    if (parts[0]!.length > 0) {
+      variants.push([parts[0]![0], ...parts.slice(1)].join(' '));
+    }
+  }
+  return [...new Set(variants)]; // deduplicate
+}
```
[the remark this answers is not in the document]
Comment (josephfung) [reply src/contacts/dedup-service.ts:112]: Declining. `nameVariants` is a private file-local helper and the noun-based name reads naturally at every call site (`nameVariants(normalized)`). Adding a `get` prefix and an alias would add indirection with no clarity gain.

File: src/contacts/dedup-service.ts [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+/**
+ * Return all blocking keys for a contact (first 3 chars of each name variant).
+ * A contact lands in multiple blocks so it can be compared against name-reversed
+ * or initial-abbreviated duplicates.
+ */
+function blockingKeys(contact: Contact): string[] {
+  const normalized = normalizeDisplayName(contact.displayName);
+  const variants = nameVariants(normalized);
+  return [...new Set(variants.map((v) => v.slice(0, 3)))];
+}
```
[the remark this answers is not in the document]
Comment (josephfung) [reply src/contacts/dedup-service.ts:123]: Declining, same reasoning as `nameVariants` above. `blockingKeys(contact)` reads clearly as a noun describing what it returns. A `getBlockingKeys` alias would add noise without benefit.

File: src/contacts/contact-service.ts [hunk 4]
```diff
@@ -175,6 +250,27 @@ export class ContactService {
...
+    if (!this.dedupService) {
+      this.logger?.warn('findDuplicates() called but no DedupService wired — returning empty');
+      return [];
+    }
+    const contacts = await this.backend.listContacts();
+    const identitiesMap = new Map<string, ChannelIdentity[]>();
+    for (const c of contacts) {
+      identitiesMap.set(c.id, await this.backend.getIdentitiesForContact(c.id));
+    }
+    return this.dedupService.findAllDuplicates(contacts, identitiesMap, minConfidence);
```
[the remark this answers is not in the document]
Comment (josephfung) [reply src/contacts/contact-service.ts]: Already done — `findDuplicates()` was wrapped in try/catch with `logger.error` before rethrow in commit `d73ffe7` (the last silent-failure review fix). The current code at this location already has the guard.

File: src/contacts/contact-service.ts [hunk 5]
```diff
@@ -418,6 +514,146 @@ export class ContactService {
...
+    const updatedPrimary: Contact = {
+      ...primary,
+      displayName: goldenRecord.displayName,
+      role: goldenRecord.role,
+      notes: goldenRecord.notes,
+      status: goldenRecord.status,
+      updatedAt: new Date(),
+    };
+    await this.backend.updateContact(updatedPrimary);
+    await this.backend.deleteContact(secondaryId);
```
[the remark this answers is not in the document]
Comment (josephfung) [reply src/contacts/contact-service.ts]: Fixed in e72261b — the merge DB write sequence (`reattachIdentities`, `reattachAuthOverrides`, `updateContact`, `deleteContact`) is now wrapped in try/catch that logs at `error` level with both contact IDs before rethrowing, so a partial-write failure is diagnosable in production.

File: src/contacts/contact-service.ts [hunk 6]
```diff
@@ -739,6 +975,48 @@ class PostgresContactBackend implements ContactServiceBackend {
...
+           SELECT channel, channel_identifier
+           FROM contact_channel_identities
+           WHERE contact_id = $2
+         )`,
+      [fromContactId, toContactId],
+    );
+    await this.pool.query(
+      `UPDATE contact_channel_identities SET contact_id = $1 WHERE contact_id = $2`,
+      [toContactId, fromContactId],
+    );
```
[the remark this answers is not in the document]
Comment (josephfung) [reply src/contacts/contact-service.ts:1005]: Declining. `reattachIdentities` is an internal backend method called from `mergeContacts()`, which now has a try/catch that logs context and rethrows (see the fix on the merge write block). Adding a second try/catch here would log a less-contextual message from the wrong level.

File: src/contacts/contact-service.ts [hunk 7]
```diff
@@ -739,6 +975,48 @@ class PostgresContactBackend implements ContactServiceBackend {
...
+           WHERE contact_id = $2 AND revoked_at IS NULL
+         )`,
+      [fromContactId, toContactId],
+    );
+    // Only re-point active (non-revoked) rows; revoked rows stay on the secondary
+    // and will be effectively deleted when the secondary contact is deleted.
+    await this.pool.query(
+      `UPDATE contact_auth_overrides SET contact_id = $1 WHERE contact_id = $2 AND revoked_at IS NULL`,
+      [toContactId, fromContactId],
+    );
```
[the remark this answers is not in the document]
Comment (josephfung) [reply src/contacts/contact-service.ts:1024]: Declining — same reasoning as the `reattachIdentities` comment above. The merge write block's try/catch covers this.

File: src/contacts/contact-service.ts [hunk 8]
```diff
@@ -739,6 +975,48 @@ class PostgresContactBackend implements ContactServiceBackend {
...
+  }
+
+  async deleteContact(id: string): Promise<void> {
+    this.logger.debug({ contactId: id }, 'contacts: deleting contact');
+    await this.pool.query(`DELETE FROM contacts WHERE id = $1`, [id]);
```
[the remark this answers is not in the document]
Comment (josephfung) [reply src/contacts/contact-service.ts:1029]: Declining — same reasoning as above. The individual backend delete method is an implementation detail; errors propagate to `mergeContacts()` where the try/catch has the context needed for a useful log message.

File: tests/unit/skills/contact-find-duplicates.test.ts [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+      contactB: { id: 'bbb', displayName: 'Bob', role: null, identities: [] },
+      score: 0.94567,
+      confidence: 'probable',
+      reason: 'Name similarity',
+    };
+    const contactService = { findDuplicates: async () => [fakePair] };
+    const result = await handler.execute(makeCtx({}, contactService));
+    if (result.success) {
+      const data = result.data as { pairs: Array<{ score: number }> };
+      expect(data.pairs[0].score).toBe(0.95);
```
[the remark this answers is not in the document]
Comment (josephfung) [reply tests/unit/skills/contact-find-duplicates.test.ts:89]: Fixed in e72261b — added `expect(result.success).toBe(true)` before the conditional block so a handler failure fails the test rather than silently passing.

File: tests/unit/contacts/dedup-service.test.ts [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+    });
+
+    it('returns probable match for similar but not identical names', () => {
+      const a = makeContact({ id: 'a', displayName: 'Jenna Torres' });
+      const b = makeContact({ id: 'b', displayName: 'Jen Torres' });
+      const results = svc.checkForDuplicates(a, [], [b], new Map());
+      // "Jenna Torres" vs "Jen Torres" — close enough to be probable or certain
+      // The exact score depends on Jaro-Winkler; just ensure we detect them
+      expect(results.length).toBeGreaterThanOrEqual(1);
+      expect(results[0].score).toBeGreaterThan(0.7);
```
[the remark this answers is not in the document]
Comment (josephfung) [reply tests/unit/contacts/dedup-service.test.ts]: Fixed in e72261b — changed `toBeGreaterThan(0.7)` to `toBeGreaterThanOrEqual(0.7)` so the threshold boundary (exactly 0.7) passes as intended.

File: tests/unit/contacts/dedup-service.test.ts [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+      const a = makeContact({ id: 'a', displayName: 'Jenna Torres' });
+      const b = makeContact({ id: 'b', displayName: 'Jenna Torres' });
+      const identitiesMap = new Map([
+        ['a', [makeIdentity('a', 'email', 'jenna@acme.com')]],
+        ['b', [makeIdentity('b', 'email', 'jenna@acme.com')]],
+      ]);
+      const allPairs = svc.findAllDuplicates([a, b], identitiesMap, 'probable');
+      const certainOnly = svc.findAllDuplicates([a, b], identitiesMap, 'certain');
+      expect(certainOnly.length).toBeLessThanOrEqual(allPairs.length);
+      expect(certainOnly.every(p => p.confidence === 'certain')).toBe(true);
```
[the remark this answers is not in the document]
Comment (josephfung) [reply tests/unit/contacts/dedup-service.test.ts:138]: The handler test only verifies the value is forwarded (`calledWith === 'certain'`) — filtering logic lives in `DedupService.findAllDuplicates()` and is unit-tested at that level (dedup-service.test.ts lines 135–138, which asserts `certainOnly.length <= allPairs.length` and every certain-only result has confidence 'certain'). The handler and the service are tested at their own contracts.

File: src/contacts/contact-service.ts [hunk 12]
```diff
@@ -1010,4 +1288,50 @@ class InMemoryContactBackend implements ContactServiceBackend {
...
+      } else {
+        this.overrides.set(id, { ...override, contactId: toContactId });
+        primaryPerms.add(override.permission);
+      }
+    }
+  }
+
+  async deleteContact(id: string): Promise<void> {
+    this.contacts.delete(id);
+  }
```
[the remark this answers is not in the document]
Comment (josephfung) [reply src/contacts/contact-service.ts:1357]: Fixed in e72261b — `InMemoryContactBackend.deleteContact()` now cascade-deletes related identities and overrides to match Postgres `ON DELETE CASCADE` and prevent test-state bleed.

File: src/contacts/dedup-service.ts [hunk 13]
```diff
@@ -0,0 +1,370 @@
...
+/**
+ * Normalize a display name for comparison:
+ * - Lowercase
+ * - Strip non-alphanumeric except spaces
+ * - Collapse whitespace
+ */
+function normalizeDisplayName(name: string): string {
+  return name
+    .toLowerCase()
+    .replace(/[^a-z0-9 ]/g, '')
```
[the remark this answers is not in the document]
Comment (josephfung) [reply src/contacts/dedup-service.ts:91]: Valid concern, noted. Added a `@TODO` comment at the strip regex in e72261b pointing to `\p{L}\p{N}` as the right fix. Changing the regex now would alter scoring for existing contacts in production and needs its own PR with migration consideration — out of scope here.
